ntoskrnl/ex/locale.c

```diff
@@ -24,8 +24,250 @@ LANGID PsInstallUILanguageId = LANGIDFROMLCID(0x00000409);
 LCID PsDefaultThreadLocaleId = 0x00000409;
 LANGID PsDefaultUILanguageId = LANGIDFROMLCID(0x00000409);
 
+/* DEFINES *******************************************************************/
+
+#define BOGUS_LOCALE_ID 0xFFFF0000
+#define LCID_HEX_FORMAT_DIGIT 8
+
 /* PRIVATE FUNCTIONS *********************************************************/
 
+/**
+ * @brief
+ * Validates the registry data of a NLS locale.
+ *
+ * @param[in] LocaleData
+ * A pointer to partial information that contains
+ * the NLS locale data.
+ *
+ * @return
+ * Returns TRUE if the following conditions are met,
+ * otherwise FALSE is returned.
+ */
+static
+BOOLEAN
+ExpValidateNlsLocaleData(
+    _In_ PKEY_VALUE_PARTIAL_INFORMATION LocaleData)
+{
+    PWCHAR Data;
+
+    /* Is this a null-terminated string type? */
+    if (LocaleData->Type != REG_SZ)
+    {
+        return FALSE;
+    }
+
+    /* Does it have a consistent length? */
+    if (LocaleData->DataLength < 2)
+    {
+        return FALSE;
+    }
+
+    /* Is the locale set and null-terminated? */
+    Data = (PWSTR)LocaleData->Data;
+    if (Data[0] != L'1' || Data[1] != UNICODE_NULL)
```

Comment:
Can you explain what that first character "1" is for?

Reply:
The "1" digit is coming from a group key from Language Groups. From my rough understanding this value data indicates that the respective language group is supported on the system. 

Any locales that is in `Locale` key but of which no language group is defined to the respective locale, the data would default to an empty value, henceforth the locale is not considered valid because it's not supported.